### DISCUSSION TITLE
How to get window handle

### DISCUSSION THREAD

**Discussion by 2gud4u** [raysan5/raylib#2973] [Q&A] [answered]
I have been trying to get a handle of the current window because i want to draw from a Dinamic or Shared Library in Linux with Raylib so far i have done this but it doesnt recognize IntPtr

`IntPtr windowHandle;
//And in main function
windowHandle = GetWindowHandle();
`
How can i get a window handle without errors ?

**Comment by jamesl-github** [comment]
are you using C# ?    
IntPtr is for C#           
found this code example 
https://gist.github.com/ChrisDill/c2ca6a0ccf592c073a8d2aaeaa4adb7f               
otherwise the cheatsheet shows it as a void * in C    
https://www.raylib.com/cheatsheet/cheatsheet.html

**Reply by 2gud4u** [reply]
Hello @jamesl-github  !!

That example was the same i found
Based on the cheatsheet then it would be like this ?
void *windowhandle;
windowhandle = GetWindowHandle();
But how do i use then the windowhandle to draw if i want to draw from a shared library for example ? I only do
BeginDrawing(); etc as normal ? I ask because i had already done it that way and it freezes the window, sorry i wish there could be more examples about this, i think its important to know this. Thank you in advance !!!

**Reply by 2gud4u** [reply]
One other thing, when i try to compile the examples from the repository i get a error in the first person camera example, the error is this.- 
`error: ‘CameraYaw’ was not declared in this scope; AND error: ‘CameraPitch’ was not declared in this scope;`
And because of that i cant implement a first person camera, @raysan5  could you help me with this ?

**Reply by raysan5** [reply]
Hi! I just tried the `core_3d_camera_first_person` example and it works ok for me. Are you using latest raylib version? Are you including `rcamera.h`?

**Comment by raysan5** [comment]
About the window handle, are you on Windows? In what platform are you trying to get it?

```c
// Get native window handle
void *GetWindowHandle(void)
{
#if defined(PLATFORM_DESKTOP) && defined(_WIN32)
    // NOTE: Returned handle is: void *HWND (windows.h)
    return glfwGetWin32Window(CORE.Window.handle);
#endif
#if defined(PLATFORM_DESKTOP) && defined(__linux__)
    // NOTE: Returned handle is: unsigned long Window (X.h)
    // typedef unsigned long XID;
    // typedef XID Window;
    //unsigned long id = (unsigned long)glfwGetX11Window(CORE.Window.handle);
    //return NULL;    // TODO: Find a way to return value... cast to void *?
    return (void *)CORE.Window.handle;
#endif
#if defined(__APPLE__)
    // NOTE: Returned handle is: (objc_object *)
    return (void *)glfwGetCocoaWindow(CORE.Window.handle);
#endif

    return NULL;
}
```

On Linux platforms, the returned window handle is probably not the expected one.

**Reply by 2gud4u** [reply]
Hello im on Linux, that explains everything, im using Raylib from the Arch Linux repos, i thought they were very recently updated, i will compile RayLib manually again, yes in the past i have included rcamera.h and still it cant find it, i will try compiling manually and i will keep you informed But if i cannot get a window handle in Linux whats the best aproach in Linux for that ?

**Reply by 2gud4u** [reply]
I have now compiled manually raylib, but it complains about not finding #include "raylib.h" :( @raysan5

**Reply by raysan5** [reply]
It seems a problem with the system include paths...

**Reply by 2gud4u** [reply]
Yes, with my old raylib from sytem repos that was fixed, but now i have remove them it cannot find raylib.h

**Reply by 2gud4u** [reply]
Oh now it works @raysan5  !!! But now how can i get the window handle in Linux ? Sorry but its extremeley important for me to know this. Did you knew my name is Jordi ? Youre from Barcelona arent you ?

**Reply by raysan5** [reply]
I'm afraid getting the window handle in Linux won't be easy, you should investigate it, I have no Linux device at hand at the moment.

**Reply by 2gud4u** [reply]
Dont want to blame you about it but thats why i always get stuck with Raylib, such a great Library/Framework but i always get stuck in key concepts, im speaking in Linux, but thank you anyway, now i will look for help and i wont find anything and ill get disapointed and drop Raylib again, yes i have leaved this before, but thank you its not your fault

**Reply by raysan5** [reply]
Why do you need it? In 9.5 years of raylib life none of +200000 users have needed it...

**Reply by raysan5** [reply]
If you really need it, maybe you should consider some lower level library like SDL, raylib is intended for higher level and that kind of very specific needs usually require some hack or internal tweak of the library to fit user needs.

**Reply by raysan5** [reply]
Also note that I'm not a Linux expert, I always work on Windows and the few times I tried using Linux my experience was quite bad. Learning how to get that Linux window handle would imply way more time for me than for a Linux experienced developer. And kind reminder that raylib is an open and collaborative project, contributions are welcome. I'm sorry if it disapoints you.

**Reply by 2gud4u** [reply]
As i said above in this thread i want to draw some 3d stuff from a shared library or dinamic library in Linux that would be the equivalent to a winows Dll but for drawing 3d stuff, yes now i understand the problem maybe its a advanced topic, and im a newbie and im asking for too much, and more for Linux, or could you tell me how can i acheive this in Linux ? please my friend ?

**Reply by 2gud4u** [reply]
No way you dont disappont me at all i have never said that

**Reply by 2gud4u** [reply]
I will close this discussion because i dont want you to feel bad, not my intention at all, you do a great job with raylib

**Reply by raysan5** [reply]
GLFW, the library used internally by raylib to manage windows exposes `glfwGetX11Window(GLFWwindow *window)` but that function requires defining `GLFW_EXPOSE_NATIVE_X11` that is unfortunately incompatible with raylib due to symbol collisions. So, I'm afraid I don't know how to get the Linux native window properly. It requires investigation and testing.

**Reply by weskerfoot** [reply]
don't mean to revive an old issue but the value being returned is an XID value which means you can pass it into Xlib or XCB functions. That works perfectly for my use case. Although if you are running Wayland it would probably return something else (not really sure what)

**Reply by weskerfoot** [reply]
actually I guess it would return NULL if you are on Wayland

**Comment by 2gud4u** [comment, marked as answer]
Linux its complex, but it should be considered in development
